TAP-2.md

```diff
@@ -37,123 +37,84 @@ Provide all on-chain liquidity in a familiar limit orderbook interface to read (
 ## Not included
-# Implementation
+# Design (draft)
 *Just a draft for inspiration – take and leave what you like.*
-```ts
-// Core Domain Types
-type Side = 'buy' | 'sell'
-type GasMode = 'include' | 'exclude'
+## Core Components
+### 1. Protocol Stream Manager
+Manages connection to Tycho Indexer for real-time protocol state updates.
+```rust
+use tycho_core::dto::Chain;
+use tycho_client::stream::TychoStreamBuilder;
 
-type TokenPair = {
-  token0: string  // Base token address
-  token1: string  // Quote token address
-}
-
-type Route = {
-  protocols: string[]  // DEX protocols used
-  calldata: string    // Executable calldata including splits
-}
-
-type Level = {
-  price: string
-  size: string
-  route: Route
-  gasCost: {
-    wei: string
-    token0: string
-    token1: string
-  }
-}
-
-// Book State
-type OrderbookState = {
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
-  bids: Level[]  // Sorted best to worst
-  asks: Level[]  // Sorted best to worst
-}
-
-// Query Options
-type OrderbookOptions = {
-  maxHops?: number
-  maxSplits?: number
-  includedProtocols?: string[]
-  pricePoints?: number
-  gasMode: GasMode
-  gasToken?: string
-  maxGasPriceGwei?: string
-}
-
-type FinetuneOptions = OrderbookOptions & {
-  tolerance: string
-  maxIterations?: number
+pub struct ProtocolStreamManager {
+    tycho_url: String,
+    auth_key: Option<String>,
+    tvl_threshold: f64,
 }
+```
+### 2. Router Interface
+An interface for routing solutions. Implement a simple solver, and expect to later integrate a more sophisticated external solver.
+```rust
+/// Generic router interface for finding and executing trades
+pub trait Router {
+    fn get_amount_out(
+        &self,
+        amount_in: U256,
+        token_in: &Address,
+        token_out: &Address,
+    ) -> Result<GetAmountOutResult, Error>;
 
-// Operation Results
-type SizeQuote = {
-  effectivePrice: string
-  route: Route
-  gasCost: Level['gasCost']
-  priceImpact: string
+    fn encode_solution(
+        &self,
+        result: GetAmountOutResult,
+        sender: Address,
+        receiver: Address,
+    ) -> Result<RoutingSolution, Error>;
 }
-
-type DepthAtPrice = {
-  totalSize: string
-  averagePrice: string
-  route: Route
-  gasCost: Level['gasCost']
+```
+### 3. Price Level Calculator 
+Calculates discrete price levels for the orderbook using router simulations.
+```rust
+pub struct PriceLevel {
+    pub price: U256,
+    pub size: U256,
+    pub solution: RoutingSolution,
 }
 
-// Event System
-type OrderbookEvent = {
-  type: 'price_level_changed'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
-  side: Side
-  price: string
-  newSize: string
-} | {
-  type: 'book_reset'
-  pair: TokenPair
-  timestamp: number
-  blockNumber: number
+pub struct PriceLevelCalculator {
+    router: Box<dyn Router>,
+    price_points: u32,
```

Comment:
Same feedback as above, this seems like it should be an internal logical component of the library and not exposed to library users - so I would not specify any code for it.